### DISCUSSION TITLE
CoreDNS Name resolution in IP4 only environment

### DISCUSSION THREAD

**Discussion by rabueker** [k3s-io/k3s#5993] [General]
I have a customer running k3s in an ipv4 only environment. Also the main DNS server (the one coredns is forwarding is requests to) resolves only ipv4. As coredns defaults to dual stack, I am getting a ServFail on the AAAA request in the pods (traced with nettools), even if the IPv4 address had been resolved properly. So the DNS lookup fails after a couple off seconds.
I was able to overcome this by adding

template ANY AAAA {
    rcode NXDOMAIN
}
to the coredns configmap, but obviously this will be overwritten in the next update.
Is there any other method outside to prevent coredns from trying to resolve IPv6 in an IP4 environment.
(net.ipv6.conf.all.disable_ipv6 = 1 on the host is not sufficient)

**Comment by brandond** [comment]
Just because you're in an ipv4-only environment does not mean that IPv6 lookups should be rejected. I suspect there is something wrong with your upstream resolver that make this necessary?

**Reply by egandro** [reply]
I have exactly that issue.

The cluster is ipv4 only (and needs to stay in that form)

Debian Node1
```
$ cat /etc/debian_version 
11.6
$ nslookup dl-cdn.alpinelinux.org
Server:         185.12.64.2
Address:        185.12.64.2#53

Non-authoritative answer:
dl-cdn.alpinelinux.org  canonical name = dualstack.d.sni.global.fastly.net.
Name:   dualstack.d.sni.global.fastly.net
Address: 146.75.122.133
Name:   dualstack.d.sni.global.fastly.net
Address: 2a04:4e42:8e::645
```

This is so stupid :/

```
$ kubectl run my-shell --rm -i --tty --image-pull-policy Always --image debian -- bash
$ apt-get update
root@my-shell:/# apt-get update
0% [Connecting to deb.debian.org (2a04:4e42:8e::644)]
```

**Reply by mysticaltech** [reply]
@brandond We are having the same issue recently! See https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/772

If you can think of any way to make a custom coredns config permanent (resisting upgrades) please let me know! 🙏

**Reply by brandond** [reply]
As I mentioned above, I don't understand why this is an issue. If your node is IPv4 only, it shouldn't be looking up AAAA records, or attempting to connect to IPv6 addresses. Do you perhaps need to take additional steps to disable IPv6 on your nodes, as hetzner has left it only partially configured in your environment?

Blocking IPv6 lookups at the coredns level is not the correct way to fix your issue.

**Reply by mysticaltech** [reply]
Ok, I understand, thanks for the tip @brandond, we will investigate at the node level more deeply. What you say makes sense!

**Reply by mysticaltech** [reply]
Ok folks, IPv6 was broken for us at the node level, as soon as we fixed the networking (just "touched" the missing NetworkManager.conf file) on our MicroOS instance, occasional IPv6 resolutions and requests started working out of the box (even though our cluster is IPv4).

I do not yet fully understand the magic, but all works, keep the good work @brandond and team! 🙏
